### PR TITLE
Codable Class Fixes

### PR DESCRIPTION
**Pull request by itaiferber**
**What's in this pull request?**
Addresses [SR-4772](https://bugs.swift.org/browse/SR-4772) and [SR-5122](https://bugs.swift.org/browse/SR-5122).

This addresses a few issues that prevented classes from properly receiving synthesized `Codable` conformances:
* #10964 allowed classes to override synthesized `Codable` conformances in subclasses; this allows a synthesized `Decodable` class to override a non-`Decodable` class by calling `super.init()` instead of `super.init(from:)` on its parent.
* Improves the diagnostics produced when a class can't receive synthesized `Decodable` implementation due to superclass limitations
* Prevents a synthesized `init(from:)` from further preventing synthesis of a default initializer (on classes) or a memberwise initializer (on structs)
* Provides improved diagnostics for suggesting that a subclass override `init(from:)` and `encode(to:)` if it could not inherit its superclass's `Codable` witnesses

rdar://problem/32774777

### CODE REVIEW

**Comment by itaiferber** [comment]
@swift-ci Please test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -2069,6 +2069,14 @@ NOTE(codable_codingkeys_type_is_not_an_enum_here,none,
      "cannot automatically synthesize %0 because 'CodingKeys' is not an enum", (Type))
 NOTE(codable_codingkeys_type_does_not_conform_here,none,
      "cannot automatically synthesize %0 because 'CodingKeys' does not conform to CodingKey", (Type))
+NOTE(decodable_no_super_init_here,none,
+     "cannot automatically synthesize %0 because superclass does not have a callable 'init()' or 'init(from:)'", (DeclName))
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: It seems like the compiler should know which of these we're trying to call, and only mention that one.

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -2069,6 +2069,14 @@ NOTE(codable_codingkeys_type_is_not_an_enum_here,none,
...
+NOTE(decodable_super_init_not_designated_here,none,
+     "cannot automatically synthesize %0 because required superclass initializer %1 is not designated", (DeclName, DeclName))
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: Hm, "required" makes me think of the keyword. Maybe just drop that word (here and below)?

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 3]
```diff
@@ -755,7 +754,7 @@ static FuncDecl *deriveEncodable_encode(TypeChecker &tc, Decl *parentDecl,
                                       TypeLoc::withoutLoc(returnType),
                                       target);
   encodeDecl->setImplicit();
-  encodeDecl->setBodySynthesizer(deriveBodyEncodable_encode);
+  encodeDecl->setBody(deriveBodyEncodable_encode(encodeDecl));
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: I think we need to continue doing this lazily. This creates extra work when just trying to use a type as Codable from another file. I think you can use `NominalTypeDecl::lookupConformance` instead.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 4]
```diff
@@ -953,47 +956,96 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
...
-                                              C.AllocateCopy(args),
-                                              C.AllocateCopy(argLabels));
+  // Classes which have a superclass must call super.init(from:) if the
+  // superclass is Decodable, or super.init() if it is not.
+  if (auto *classDecl = dyn_cast<ClassDecl>(targetDecl)) {
+    if (auto *superclassDecl = classDecl->getSuperclassDecl()) {
+      auto superType =  superclassDecl->getDeclaredInterfaceType();
+      if (tc.conformsToProtocol(superType,
+                                C.getProtocol(KnownProtocolKind::Decodable),
+                                superclassDecl, ConformanceCheckFlags::Used)) {
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: This doesn't actually need to be `Used`, since nothing in this body depends on the conformance.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 5]
```diff
@@ -953,47 +956,96 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
...
+        auto *superDecoderCall =
+          CallExpr::createImplicit(C, superDecoderRef, ArrayRef<Expr *>(),
+                                   ArrayRef<Identifier>());
+
+        // super
+        auto *superRef = new (C) SuperRefExpr(initDecl->getImplicitSelfDecl(),
+                                              SourceLoc(), /*Implicit=*/true);
+
+        // super.init(from:)
+        auto initName = DeclName(C, C.Id_init, (Identifier[1]){ C.Id_from });
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: There's an implicit conversion from a single element to an ArrayRef, so you can avoid the inline array literal.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 6]
```diff
@@ -953,47 +956,96 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
...
+                                        /*Implicit=*/true);
+        statements.push_back(tryExpr);
+      } else {
+        // The explicit constructor name is a compound name taking no arguments.
+        DeclName initName(C, C.Id_init, ArrayRef<Identifier>());
+
+        // We need to look this up in the superclass to see if it throws.
+        // We should have bailed one level up if super.init() is not available.
+        ConstructorDecl *superInitDecl = nullptr;
+        for (auto *decl : superclassDecl->getMembers()) {
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: Please use `lookupDirect` instead of walking the members.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 7]
```diff
@@ -953,47 +956,96 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
...
+        // super.init()
+        auto *superInitRef = new (C) UnresolvedDotExpr(superRef, SourceLoc(),
+                                                       initName, DeclNameLoc(),
+                                                       /*Implicit=*/true);
+        // super.init() call
+        auto *callExpr = CallExpr::createImplicit(C, superInitRef,
+                                                  ArrayRef<Expr *>(),
+                                                  ArrayRef<Identifier>());
+
+        if (superInitDecl->hasThrows()) {
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: What if it's failable?

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 8]
```diff
@@ -1107,10 +1159,73 @@ static ValueDecl *deriveDecodable_init(TypeChecker &tc, Decl *parentDecl,
...
+  // Before we attempt to look up (or more importantly, synthesize) a CodingKeys
+  // entity on target, we need to make sure the type is otherwise valid.
+  //
+  // If we are synthesizing Decodable and the target is a class with a
+  // superclass, our synthesized init(from:) will need to call either
+  // super.init(from:) or super.init() depending on whether the superclass is
+  // Decodable itself.
+  //
+  // If the required initializer is not available, we shouldn't attempt to
+  // synthesize CodingKeys.
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: Won't it be a problem if the subclass and the superclass both have a CodingKeys enum?

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 9]
```diff
@@ -1107,10 +1159,73 @@ static ValueDecl *deriveDecodable_init(TypeChecker &tc, Decl *parentDecl,
...
+  ASTContext &C = tc.Context;
+  auto *classDecl = dyn_cast<ClassDecl>(target);
+  if (proto == C.getProtocol(KnownProtocolKind::Decodable) && classDecl) {
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: Please use `ProtocolDecl::isSpecificProtocol` instead.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 10]
```diff
@@ -1107,10 +1159,73 @@ static ValueDecl *deriveDecodable_init(TypeChecker &tc, Decl *parentDecl,
...
+    if (auto *superclassDecl = classDecl->getSuperclassDecl()) {
+      DeclName memberName;
+      auto superType = superclassDecl->getDeclaredInterfaceType();
+      if (tc.conformsToProtocol(superType, proto, superclassDecl,
+                                ConformanceCheckFlags::Used)) {
+        // super.init(from:) must be accessible.
+        memberName = cast<ConstructorDecl>(requirement)->getEffectiveFullName();
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: This should just be `getFullName()`.

**Review by jrose-apple — COMMENTED** [review]
Glad you're tackling this! Some early comments, and also a request to add more tests, particularly around what happens if the superclass initializer fails.

**Comment by itaiferber** [comment]
@jrose-apple Thanks for the review! Looking through your comments now.

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (itaiferber) [inline include/swift/AST/DiagnosticsSema.def]: I had this show the specific member directly in an earlier version of this, but wanted to highlight that you can avoid this issue by either making `super` `Decodable`, or provide an `init()` (regardless of which case you were currently in). Do we prefer to just show the one we need?

File: include/swift/AST/DiagnosticsSema.def (hunk 2 above)
Comment (itaiferber) [inline include/swift/AST/DiagnosticsSema.def]: To go along with the comment on the other diagnostics — what I'm trying to get across here (perhaps poorly) is that the implementation of `init(from:)` we'd generate needs to be able to call this member. Trying to explain this in a diagnostic gets a bit wordy, so I settled on "required" because it appeared in other diagnostics.

Another option I tried (but opted not to use) was `cannot synthesize %0 because implementation would need to call %1, which is not designated` (or similar). Does that sound more reasonable, or should I find a different way to rephrase?

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: I think so. You don't usually have the option to change your superclass, right?

File: include/swift/AST/DiagnosticsSema.def (hunk 2 above)
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: That seems fine. Ultimately we should be optimizing for easy-to-understand, here.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 7 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: Good point! Is there any way you're allowed to call an optional initializer on super from a non-optional initializer in any way other than `super.init()!`? Seems like you can't put `super.init()` in any statements.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 8 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: No — they should both be `private`. We will never synthesize a non-private `CodingKeys` enum

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (itaiferber) [inline include/swift/AST/DiagnosticsSema.def]: True. I guess we can change this later if it seems necessary.

File: include/swift/AST/DiagnosticsSema.def (hunk 2 above)
Comment (itaiferber) [inline include/swift/AST/DiagnosticsSema.def]: I completely agree — I wanted to hit a spot somewhere between informative and "fits in a reasonable Xcode window" 😅

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 7 above)
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: Yeah, I think that's the only way. I'm just not sure how much you/we care about these cases at all.

Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: I think it's reasonable to call it rather than bail if it's failable.

Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: A thought: it would be safer to bail and add the calling later if people complain rather than the other way around.

Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: Fair, I'll do that instead.

**Comment by itaiferber** [comment]
@swift-ci Please test

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 11]
```diff
@@ -344,8 +344,9 @@ static bool canSynthesizeRawRepresentable(TypeChecker &tc, Decl *parentDecl,
   auto parentDC = cast<DeclContext>(parentDecl);
   rawType       = parentDC->mapTypeIntoContext(rawType);
 
-  if (!enumDecl->getInherited().empty() &&
-      enumDecl->getInherited().front().isError())
+  auto inherited = enumDecl->getInherited();
+  if (!inherited.empty() && inherited.front().wasValidated() &&
+      inherited.front().isError())
```
Comment (itaiferber) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: It turns out that the `wasValidated()` check is indeed necessary. If we hit this _during_ validation of the `enumDecl`, and while the raw type is already being validated (but hasn't finished yet), `isError()` will assert.

File: test/decl/protocol/special/coding/class_codable_failure_diagnostics.swift [hunk 12]
```diff
@@ -8,12 +8,12 @@ class C1 : Codable {
   var c: Nested = Nested()
 
   // CHECK: error: type 'C1' does not conform to protocol 'Decodable'
-  // CHECK: note: cannot automatically synthesize 'Decodable' because 'C1.Nested' does not conform to 'Decodable'
```
Comment (itaiferber) [inline test/decl/protocol/special/coding/class_codable_failure_diagnostics.swift]: The ordering of these diagnostics has changed.

File: test/decl/protocol/special/coding/struct_codable_failure_diagnostics.swift [hunk 13]
```diff
@@ -6,14 +6,6 @@ struct S1 : Codable {
   var a: String = ""
   var b: Int = 0
   var c: Nested = Nested()
-
-  // CHECK: error: type 'S1' does not conform to protocol 'Decodable'
```
Comment (itaiferber) [inline test/decl/protocol/special/coding/struct_codable_failure_diagnostics.swift]: The ordering of these diagnostics has changed. Since we perform `Decodable` conformance checking in `addImplicitConstructors`, we hit all `Decodable` conformances first, and `Encodable` checking happens afterwards. These `CHECK`s can no longer be interleaved, so they're at the bottom of the file.

**Comment by itaiferber** [comment]
@jrose-apple I believe I've addressed most if not all of your concerns with this last push. Happy to receive any other comments or feedback. 🙂

**Comment by itaiferber** [comment]
Looks like the result type of `TypeChecker::lookupMember` has changed on `master`. Fixing now.

**Comment by itaiferber** [comment]
@swift-ci Please test

File: lib/Sema/DerivedConformanceRawRepresentable.cpp (hunk 11 above)
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: That's a problem too, right? If we bail out here, will we just not synthesize anything?

Comment (itaiferber) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: That's correct. Though, the `RawType` here was an error type anyway — there's no way to recover from that...
This code already exists and hasn't been an issue; I just want this to stop asserting incorrectly. If we'd prefer to take it out, I don't mind.

Comment (itaiferber) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: In fact, I think we've got some tests which test failure to synthesize `RawRepresentable` for `@objc` enums where the raw type is invalid. I can take a look.

**Comment by itaiferber** [comment]
Looks like the macOS builder's been stuck since this morning on one of the `watchOS` tests: `09:15:30 99%: XcodeprojTests.FunctionalTests/testXcodeProjWithPkgConfig`

Going to attempt to re-build.

**Comment by itaiferber** [comment]
@swift-ci Please clean test macOS

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 14]
```diff
@@ -944,56 +957,99 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
...
+        // The explicit constructor name is a compound name taking no arguments.
+        DeclName initName(C, C.Id_init, ArrayRef<Identifier>());
+
+        // We need to look this up in the superclass to see if it throws.
+        ConstructorDecl *superInitDecl = nullptr;
+        auto result = superclassDecl->lookupDirect(initName);
+
+        // We should have bailed one level up if this were not available.
+        assert(!result.empty() &&
+               (superInitDecl = dyn_cast<ConstructorDecl>(result.front())));
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: You can't put side effects into an `assert`; they'll be compiled out in Release builds.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 15]
```diff
@@ -944,56 +957,99 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
...
+                                                       initName, DeclNameLoc(),
+                                                       /*Implicit=*/true);
+        // super.init() call
+        Expr *callExpr = CallExpr::createImplicit(C, superInitRef,
+                                                  ArrayRef<Expr *>(),
+                                                  ArrayRef<Identifier>());
+
+        // If super.init is failable, super.init()!
+        if (superInitDecl->getFailability() != OTK_None)
+          callExpr = new (C) ForceValueExpr(callExpr, SourceLoc());
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: I thought you decided to complain about this?

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 16]
```diff
@@ -1053,14 +1109,12 @@ static ValueDecl *deriveDecodable_init(TypeChecker &tc, Decl *parentDecl,
...
-      /*FailabilityLoc=*/SourceLoc(),
-      /*Throws=*/true, /*ThrowsLoc=*/SourceLoc(), selfDecl, paramList,
-      /*GenericParams=*/nullptr, target);
+  auto *initDecl = new (C) ConstructorDecl(name, SourceLoc(), OTK_None,
+                                           SourceLoc(), /*Throws=*/true,
+                                           SourceLoc(), selfDecl, paramList,
+                                           /*GenericParams=*/nullptr, target);
   initDecl->setImplicit();
-  initDecl->setBodySynthesizer(deriveBodyDecodable_init);
+  initDecl->setBody(deriveBodyDecodable_init(tc, initDecl));
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: Again, please do not synthesize the body eagerly.

File: lib/Sema/TypeCheckDecl.cpp [hunk 17]
```diff
@@ -7946,6 +7946,67 @@ static void diagnoseClassWithoutInitializers(TypeChecker &tc,
...
+  //
+  // It is helpful to suggest here that the user may have forgotten to override
+  // init(from:) (and encode(to:), if applicable) in a note, before we start
+  // listing the members that prevented initializer synthesis.
+  // TODO: Add a fixit along with this suggestion.
+  if (auto *superclassDecl = classDecl->getSuperclassDecl()) {
+    ASTContext &C = tc.Context;
+    auto *decodableProto = C.getProtocol(KnownProtocolKind::Decodable);
+    auto superclassType = superclassDecl->getDeclaredInterfaceType();
+    if (auto ref = tc. conformsToProtocol(superclassType, decodableProto,
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: Stray space snuck in here.

File: lib/Sema/TypeCheckDecl.cpp [hunk 18]
```diff
@@ -7946,6 +7946,67 @@ static void diagnoseClassWithoutInitializers(TypeChecker &tc,
...
+                                          superclassDecl,
+                                          ConformanceCheckOptions(),
+                                          SourceLoc())) {
+      // super conforms to Decodable, so we've failed to inherit init(from:).
+      // Let's suggest overriding it here.
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: Hm. I'm not sure whether this should error if there are no stored properties on the subclass. It doesn't matter up until someone adds a stored property, and now suddenly existing archives are missing an extra level of "super" container and they have to deal with that specially.

File: lib/Sema/TypeCheckDecl.cpp [hunk 19]
```diff
@@ -8133,14 +8194,48 @@ void TypeChecker::addImplicitConstructors(NominalTypeDecl *decl) {
...
+  // initializers are properly synthesized.
+  //
+  // NOTE: Lookups of synthesized initializers MUST come after
+  //       decl->setAddedImplicitInitializers() in case synthesis requires
+  //       protocol conformance checking, which might be recursive here.
+  DeclName synthesizedInitializers[1] = {
+    // init(from:) is synthesized by derived conformance to Decodable.
+    DeclName(Context, DeclBaseName(Context.Id_init),
+             (Identifier[1]) { Context.Id_from })
+  };
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: I'm not sure this matters. No implicit initializer ought to prevent initializer synthesis or inheritance.

File: test/decl/protocol/special/coding/struct_codable_memberwise_initializer.swift [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+  }
+}
+
+// The same should hold for structs whose members all have initial values.
+struct InitialValueNoInitializers {
+  var x: Double = .pi
+
+  func foo() {
+    // The struct should receive a memberwise initializer.
+    let _ = NoInitializers.init(x:)
```
Comment (jrose-apple) [inline test/decl/protocol/special/coding/struct_codable_memberwise_initializer.swift]: …and a no-argument initializer.

File: test/decl/protocol/special/coding/class_codable_default_initializer.swift [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+class CodableNoExplicitInitializers : Codable {
+  var x: Double
+
+  func foo() {
+    // The class should receive a synthesized init(from:) and encode(to:)
+    let _ = CodableNoExplicitInitializers.init(from:)
+    let _ = CodableNoExplicitInitializers.encode(to:)
+
+    // It should not, however, receive a default constructor.
+    let _ = CodableNoExplicitInitializers.init
```
Comment (jrose-apple) [inline test/decl/protocol/special/coding/class_codable_default_initializer.swift]: Checking for an ambiguity here is a little subtle. Maybe just try calling it, like you do below?

**Review by jrose-apple — COMMENTED** [review]
Still more comments, sorry.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 14 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: Wow, you're right — don't know how I missed that. Will fix.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 15 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: We do (and bail) one level up — we should never hit this, and I can remove it (or replace with an assert that we should never hit this).

Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: Assertion sounds good for now. As long as there's a test case for it, it'll be caught.

…hm. You're missing a few test cases.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 16 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: Unfortunately, I think this is unavoidable here. We need to pass the `TypeChecker` in for member lookups, and the lazy synthesizer version only takes functions which have one argument — the decl itself. We've run into this before, I think. The alternative is to get an `ASTContext` from the `initDecl` and cast its lazy resolver to a `TypeChecker`, but I'm not sure how safe that is...

File: lib/Sema/TypeCheckDecl.cpp (hunk 17 above)
Comment (itaiferber) [inline lib/Sema/TypeCheckDecl.cpp]: Good catch, thanks!

File: lib/Sema/TypeCheckDecl.cpp (hunk 19 above)
Comment (itaiferber) [inline lib/Sema/TypeCheckDecl.cpp]: Unfortunately, we do have cases where we appear to rely on this (I believe somewhere deep in the `String` implementation in the stdlib). The first thing I tried was simply to prevent any implicit initializer from preventing synthesis, but the `stdlib` couldn't compile because it was crashing in compiling `String` (or one of the related types).

I can investigate this further to give you more info, but this was to avoid that from happening. ☹️

File: test/decl/protocol/special/coding/class_codable_default_initializer.swift (hunk 21 above)
Comment (itaiferber) [inline test/decl/protocol/special/coding/class_codable_default_initializer.swift]: Can do! 👍

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 16 above)
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: That's acceptable, but shouldn't you be able to use `DeclContext::lookupQualified` for lookups?

File: lib/Sema/TypeCheckDecl.cpp (hunk 19 above)
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: Hm, we should investigate later, but it doesn't have to block you for now. Can you put a FIXME, though?

Comment (itaiferber) [inline lib/Sema/TypeCheckDecl.cpp]: Can do!

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 16 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: If `super.init()` is synthesized, `DeclContext::lookupQualified` won't always be able to it to be synthesized unless we call `TypeChecker::addImplicitConstructors` from `TypeChecker::synthesizeMemberForLookup` (whereas `tc.lookupMember` does synthesize it if need be). The one danger to adding the `addImplicitConstructors` call is accidentally making `synthesizeMemberForLookup` recursive.

We also use the type checker to see if super conforms to `Decodable`, but we might be able to avoid that.

Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: It already does that; that's what the call to `resolveImplicitConstructors` is for.

Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: I thought there was something preventing this, but let me take a look — if I can avoid using the `TypeChecker`, it would certainly be nice to.

**Comment by itaiferber** [comment]
@swift-ci Please test

File: lib/Sema/TypeCheckDecl.cpp (hunk 18 above)
Comment (itaiferber) [inline lib/Sema/TypeCheckDecl.cpp]: You're right, I'll fix this.

**Comment by itaiferber** [comment]
@jrose-apple Another push with fixes. Thanks for reviewing this so thoroughly! I appreciate it. Happy to fix anything else you don't feel confident in.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 22]
```diff
@@ -944,55 +953,95 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
...
+
+        statements.push_back(callExpr);
+      }
+    }
   }
 
-  auto *body = BraceStmt::create(C, SourceLoc(), statements, SourceLoc(),
-                                 /*implicit=*/true);
+  auto *body =  BraceStmt::create(C, SourceLoc(), statements, SourceLoc(),
+                                  /*implicit=*/true);
```
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: Extra space here too, for some reason.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 8 above)
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceCodable.cpp]: What if somebody *else* declares a non-private CodingKeys enum on the superclass?

File: lib/Sema/TypeCheckDecl.cpp (hunk 18 above)
Comment (itaiferber) [inline lib/Sema/TypeCheckDecl.cpp]: Actually, we never hit this diagnostic in cases like this. If the class adds no stored properties (`SuppressDefaultInitializer` in `addImplicitConstructors` is false), we always try to inherit the superclass's initializers. If it has none (because synthesis failed or for any other reason), we simply don't inherit any, and never complain.

This is true in Swift 3 and Swift 4:

```swift
class Super {
	var value: Double
}

class Sub : Super {}
```

If we'd like to complain there, then I can add a check in both locations (whether we have no initializers and also inherited none from `super`, and also whether we're about to produce a diagnostic about something that's not correct); otherwise we're safe to leave things as-is.

**Review by jrose-apple — COMMENTED** [review]
I *think* this is all I have!

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 22 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: 😒 vim

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 8 above)
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: @jrose-apple I don't think that's problematic — we can always add a private type to a subclass with the same name as a superclass's public type:

```swift
public class Super {
    public enum X {
        case somePublic
    }
}

public class Sub : Super {
    private enum X {
        case myPrivate
    }
}
```

^ works just fine
We always synthesize a private enum, so I don't think this is problematic.

**Comment by itaiferber** [comment]
@jrose-apple If this looks good to you at this point, I'd like to get this merged today. I'm going to retest here to make sure that things haven't changed on master that would break the build to be safe.

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@DougGregor is back now, but wading through piles of email and PRs. If he hasn't commented by this afternoon then I say go for it.

**Comment by itaiferber** [comment]
@DougGregor I know you're really busy catching up with stuff but I'd like to get this merged in within the next hour or so so I have time to put this up against `swift-4.0-branch`. If there's anything prohibitive about this PR, do let me know — otherwise I'm happy to make small fixes in future PRs.
